### PR TITLE
Qualcomm AI Engine Direct - retire "_unlift" IR capture option

### PR DESCRIPTION
**Pull request by haowhsu-quic**
- Remove "_unlift" from qnn_capture_config
- Refactor file architecture and functions

### CODE REVIEW

**Comment by haowhsu-quic** [comment]
Hi @kimishpatel,

This PR removes `_unlift` config option, please kindly have a look. (The CI lint error seems not to be introduced by this change, should I need to fix it?)

About Qualcomm backend CI job, we probably need to wait for the ongoing discussion about artifacts location to be done. Before that, we'll make sure everything works with latest mainline locally.
We also plan to have PRs refactoring or enhancing current implementation in the near future. (quantizer, runtime performance control..., etc.)

Thank you.

**Comment by kimishpatel** [comment]
> About Qualcomm backend CI job, we probably need to wait for the ongoing discussion about artifacts location to be done. Before that, we'll make sure everything works with latest mainline locally.

@haowhsu-quic thank you very much. will take a look

File: backends/qualcomm/builders/node_visitor.py [hunk 1]
```diff
@@ -49,51 +51,37 @@ class NodeVisitor:
...
         """
-        Get tensor shape with axis_order
+        Get tensor value/shape with axis_order
         """
 
-        def get_tensor(node, index):
+        def _get_tensor(node, index):
             if index is not None:
                 assert isinstance(index, int)
+                if is_parameter(node, self.edge_program):
```
Comment (kimishpatel) [inline backends/qualcomm/builders/node_visitor.py:69]: @angelayi @tugsbayasgalan can we introduce utils on the ExportedProgram to check if a node is parameter or buffer node and then have utils get_parameter?

Basically
`EdgeProragram.is_parameter(node)` `EdgeProgram.get_parameter(node)`

File: backends/qualcomm/partition/qnn_partitioner.py [hunk 2]
```diff
@@ -41,25 +39,20 @@ def __init__(self, graph_module: torch.fx.GraphModule, compiler_specs):
...
 
-        if op_wrapper is None:
-            print(f"[QNN Partitioner Op Support]: {node.target.__name__} | False")
-            return False
-
-        supported = self.qnn_manager.IsNodeSupportedByBackend(
-            [op_wrapper.GetOpWrapper()]
+        supported = False
+        op_wrapper = self.node_visitors[node.target.__name__].define_node(
+            node, self.nodes_to_wrappers
```
Comment (kimishpatel) [inline backends/qualcomm/partition/qnn_partitioner.py:50]: nit: it feels nodes_to_wrapers should really be owned by node_visitor? And if it should be owned by partitioner, I would imagine then all the caching logic should be inside partitioner?

File: backends/qualcomm/passes/annotate_and_quant_scalar.py [hunk 3]
```diff
@@ -37,24 +38,22 @@ class AnnotateAndQuantScalar(ExportPass):
...
+    def __init__(self, edge_program: torch.export.ExportedProgram):
         super(AnnotateAndQuantScalar, self).__init__()
+        self.edge_program = edge_program
 
     def _get_source_scalar_node(self, node: torch.fx.Node) -> torch.fx.Node:
         """
         This recursion function is specific for multiply followed by a cast
         """
-        if node.op == "get_attr":
+        if node.op == "placeholder":
```
Comment (kimishpatel) [inline backends/qualcomm/passes/annotate_and_quant_scalar.py:49]: Scalar arguments should really be "burnt" into the node and I dont expect them to be placholder (and get_attr) nodes. Is that not what you see?

File: backends/qualcomm/passes/recompose_pixel_shuffle.py [hunk 4]
```diff
@@ -1,69 +0,0 @@
-# Copyright (c) Qualcomm Innovation Center, Inc.
```
Comment (kimishpatel) [inline backends/qualcomm/passes/recompose_pixel_shuffle.py:1]: This isnt needed anymore?

File: backends/qualcomm/passes/utils.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    quant_attr_keys = [arg.name for arg in quant_node.target._schema.arguments][1:]
+    quant_attrs = dict.fromkeys(quant_attr_keys)
+
+    for i in range(1, len(quant_node.args)):
+        attr_n = quant_node.args[i]
+
+        value = attr_n
+        if type(attr_n) == torch.fx.node.Node:
+            # could be a commonly shared attribute between q & dq
+            if attr_n.target == exir_ops.edge.aten._to_copy.default:
```
Comment (kimishpatel) [inline backends/qualcomm/passes/utils.py:24]: Why this special handing. Are some quant node's args are from copy? and why?

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 6]
```diff
@@ -121,22 +121,6 @@ def forward(self, x):
         model_name = "qnn_mean_dim_model"
         save_model_and_expected_output(instance, buffer, example_inputs, model_name)
 
-    def test_qnn_backend_addmm(self):
```
Comment (kimishpatel) [inline backends/qualcomm/tests/test_qnn_delegate.py:124]: why is this test removed?

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 7]
```diff
@@ -561,7 +543,7 @@ def __init__(self):
                 self.avgPool = torch.nn.AvgPool2d(
                     kernel_size=(2, 2),
                     padding=(1, 1),
-                    stride=(2, 2),
+                    stride=(1, 1),
```
Comment (kimishpatel) [inline backends/qualcomm/tests/test_qnn_delegate.py:546]: why this change

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 8]
```diff
@@ -473,9 +457,6 @@ def forward(self, x, y):
                 return self.sub(x, y)
 
         instance = Sub()
-        example_inputs = (torch.randn(2, 5, 1, 3), torch.randn(2, 5, 1, 3))
-        self.lower_module_and_test_output(instance, example_inputs)
```
Comment (kimishpatel) [inline backends/qualcomm/tests/test_qnn_delegate.py:477]: why is this removed?

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 9]
```diff
@@ -1126,31 +1104,6 @@ def forward(self, x):
         model_name = "ptq_qnn_pixel_shuffle_model"
         save_model_and_expected_output(instance, buffer, example_inputs, model_name)
 
-    def test_qnn_backend_ptq_home_made_pixel_shuffle(self):
```
Comment (kimishpatel) [inline backends/qualcomm/tests/test_qnn_delegate.py:1129]: same here. Why is this test removed?

File: backends/qualcomm/utils/utils.py [hunk 10]
```diff
@@ -37,12 +50,28 @@ def capture_program(
...
         qnn_capture_config(),
     ).to_edge(qnn_edge_config())
-    return edge_prog.transform(qnn_partitioner_passes)
+
+    # currently ExirExportedProgram.transform does not accept
+    # changes of input number which was caused by FoldQDQ
+    # apply passes one by one here to avoid IR capture failure
+    edge_program = ex_prog.exported_program
+    graph_module = edge_program.graph_module
+    AddmmToLinearTransform()(graph_module)
```
Comment (kimishpatel) [inline backends/qualcomm/utils/utils.py:64]: You can also use this https://github.com/pytorch/pytorch/blob/main/torch/export/exported_program.py#L480 to apply transformation passes.

**Review by kimishpatel — APPROVED** [review]
Largely looks good. Can you also have someone from you team review since I wont be paying too close attention to the core of your code.

For future though, please do separate changes as appropriate to keep PR small. For now just accepting it.

File: backends/qualcomm/partition/qnn_partitioner.py (hunk 2 above)
Comment (haowhsu-quic) [reply backends/qualcomm/partition/qnn_partitioner.py:50]: Thank you for the idea. Since `node_visitor` is a dictionary, if `nodes_to_wrappers` was owned by each instance, we might need to traverse all for them to get the cached tensors.

File: backends/qualcomm/passes/annotate_and_quant_scalar.py (hunk 3 above)
Comment (haowhsu-quic) [reply backends/qualcomm/passes/annotate_and_quant_scalar.py:49]: No, it will be translated into placeholder after IR capture. The lifting part is introduce by https://github.com/pytorch/pytorch/blob/main/torch/_export/passes/lift_constant_tensor_pass.py

File: backends/qualcomm/passes/fold_qdq.py [hunk 11]
```diff
@@ -56,6 +56,9 @@ def _fold(self, graph_module: torch.fx.GraphModule) -> torch.fx.GraphModule:
             for i in range(1, len(n.args)):
                 if type(n.args[i]) == torch.fx.node.Node:
                     to_be_removed.append(n.args[i])
+                    # could be a commonly shared attribute between q & dq
+                    if n.args[i].target == exir_ops.edge.aten._to_copy.default:
+                        to_be_removed.append(n.args[i].args[0])
```
[the remark this answers is not in the document]
Comment (haowhsu-quic) [reply backends/qualcomm/passes/fold_qdq.py:61]: This part is necessary, since the q / dq node of parameters in placeholder format will commonly share the same scale and offset. We need to tackle it with this approach to correctly fill in quantization attribute.

File: backends/qualcomm/passes/recompose_pixel_shuffle.py (hunk 4 above)
Comment (haowhsu-quic) [reply backends/qualcomm/passes/recompose_pixel_shuffle.py:1]: This test case was first introduced by EDSR where existed a mathematically equivalent pixel shuffle implemented with small operators. It seems to be changed back to `nn.PixelShuffle` in the latest example.

File: backends/qualcomm/passes/utils.py (hunk 5 above)
Comment (haowhsu-quic) [reply backends/qualcomm/passes/utils.py:24]: This is also introduced by lifting behavior, the parameters in placeholder format will share the same scale and offset for the consecutive q and dq node.

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 6 above)
Comment (haowhsu-quic) [reply backends/qualcomm/tests/test_qnn_delegate.py:124]: I think we don't really have an `addmm` operator in existing example models, and there also has no builder for it.
We will add it back once the use case was found.

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 7 above)
Comment (haowhsu-quic) [reply backends/qualcomm/tests/test_qnn_delegate.py:546]: We notice this test parameter is break in some version of QNN. Change to a simpler case to make unit test focus on the framework, not the SDK itself.

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 8 above)
Comment (haowhsu-quic) [reply backends/qualcomm/tests/test_qnn_delegate.py:477]: Both test cases were work, remove it here because we hope each test case produces only one artifact, for the ease of our local testing.
We will make the test as comprehensive as possible when we start doing CI integration, thank you.

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 9 above)
Comment (haowhsu-quic) [reply backends/qualcomm/tests/test_qnn_delegate.py:1129]: The reason is similar to the change of EDSR model as aforementioned, thanks.

**Comment by haowhsu-quic** [comment]
Thank you very much, will pay attention to the rule more carefully.

**Review by shewu-quic — APPROVED** [review]
Look good for me.

**Review by chunit-quic — APPROVED** [review]
LGTM

File: backends/qualcomm/builders/node_visitor.py (hunk 1 above)
Comment (angelayi) [reply backends/qualcomm/builders/node_visitor.py:68]: I think we already have [these helper functions](https://github.com/pytorch/pytorch/blob/e39668770a9bb38379803845a13e3f0069db62ea/torch/_export/utils.py#L97), but maybe we can add to the exported program too.
